modules/local_contacts/modules.php

```diff
@@ -65,6 +65,92 @@ public function process() {
+        if ($success && $form['contact_source'] == 'csv') {
+            $file = $this->request->files['contact_csv'];
+            $csv = fopen($file['tmp_name'], 'r');
+            if ($csv) {
+                $contacts = $this->get('contact_store');
+                $header = fgetcsv($csv);
+                $expectedHeader = array('display_name', 'email_address', 'phone_number');
+
+                if ($header !== $expectedHeader) {
+                    fclose($csv);
+                    Hm_Msgs::add('ERRInvalid CSV file, please use a valid header: '.implode(', ', $expectedHeader));
+                    return;
+                }
+
+                $contact_list = $this->user_config->get('contacts', array());
+                $contact_list = array_map(function($v) { $v['type'] = 'local'; return $v; }, $contact_list);
+                $message = '';
+                $update_count = 0;
+                $create_count = 0;
+                $invalid_mail_count = 0;
+                $import_result = [];
+                
+
+                while (($data = fgetcsv($csv)) !== FALSE) {
+                    $single_contact = [
+                        'display_name' => $data[0],
+                        'email_address' => $data[1],
+                        'phone_number' => $data[2] ?? ''
+                    ];
+                    $email = $data[1];
+                    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+                        $single_contact['status'] = 'invalid email';
+                        array_push($import_result, $single_contact);
+                        $invalid_mail_count++;
+                        continue;
+                    }
+
+                    $details = array('source' => 'local', 'display_name' => $data[0], 'email_address' => $email);
+                    if (array_key_exists(2, $data) && $data[2]) {
+                        $details['phone_number'] = $data[2];
+                    }
+
+                    $contactUpdated = false;
+
+                    foreach ($contact_list as $key => $contact) {
+                        if ($contact['email_address'] == $email) {
+                            $contacts->update_contact($key, $details);
+                            $single_contact['status'] = 'update';
+                            array_push($import_result, $single_contact);
+                            $update_count++;
+                            $contactUpdated = true;
+                            continue 2;
+                        }
+                    }
+
+                    if (!$contactUpdated) {
+                        $contacts->add_contact($details);
+                        $single_contact['status'] = 'new';
+                        array_push($import_result, $single_contact);
+                        $create_count++;
+                    }
+                }
+                fclose($csv);
+                $this->user_config->set('contacts', $contacts->export());
```

Comment:
There is also a new way to save the changes. Please check other contact modules.